dask/bytes/core.py

```diff
@@ -149,6 +231,107 @@ def open_files(urlpath, compression=None, **kwargs):
+                     "name_function must preserve the order of its input")
+
+            paths = [path.replace('*', name_function(i))
+                     for i in range(num)]
+        else:
+            paths = [path]
+    elif isinstance(path, (tuple, list, set)):
+        assert len(path) == num
+        paths = path
+    else:
+        raise ValueError("""Path should be either"
+1.  A list of paths -- ['foo.json', 'bar.json', ...]
+2.  A directory -- 'foo/
+3.  A path with a * in it -- 'foo.*.json'""")
+    return paths
+
+
+def write_files(nfiles, urlpath, name_function=None, compression=None,
+                lazy=False, mode='wb', **kwargs):
+    """ Given path return dask.delayed file-like objects for writing
+
+    Parameters
+    ----------
+    values: list/set of values or collection
+        things that return bytes to be written into files
+    urlpath: string
+        Absolute or relative filepath, URL (may include protocols like
+        ``s3://``), or globstring pointing to data output.
+    name_function: function (None)
+        Way to convert each part into an identifier, if using a url globstring,
+        default: dask.utils.build_name_function
+    compression: string (None)
+        Compression to use.  See ``dask.bytes.compression.files`` for options.
+    lazy: bool (True)
+        whether to perform the write immediately, or return values which write
+        upon evaluation
+    **kwargs: dict
+        Extra options that make sense to a particular storage connection, e.g.
+        host, port, username, password, etc.
+
+    Examples
+    --------
+    >>> files = write_files('2015-*.csv')  # doctest: +SKIP
+    >>> files = open_files('s3://bucket/2015-*.csv.gz', compression='gzip')  # doctest: +SKIP
+
+    Returns
+    -------
+    List of ``dask.delayed`` objects that compute to file-like objects
+    """
+    if compression == 'infer':
+        if isinstance(urlpath, (tuple, list, set)):
+            compression = infer_compression(urlpath[0])
+        else:
+            compression = infer_compression(urlpath)
+    if compression is not None and compression not in compress_files:
+        raise ValueError("Compression type %s not supported" % compression)
+
+    storage_options = infer_storage_options(urlpath,
+                                            inherit_storage_options=kwargs)
+    paths = _expand_paths(urlpath, name_function, nfiles)
+
+    protocol = storage_options.pop('protocol')
+    ensure_protocol(protocol)
+    try:
+        open_files_backend = _write_files[protocol]
+    except KeyError:
+        raise NotImplementedError("Unknown protocol %s (%s)" %
+                                  (protocol, urlpath))
+
+    # backend file is always binary
+    mode2 = mode.replace('t', '').replace('b', '') + 'b'
+    files = open_files_backend(paths, mode=mode2, **kwargs)
+    if 'b' not in mode:
+        # wrap for text mode
+        files = [delayed(lambda f: io.TextIOWrapper(SeekableFile(f)))(f)
```

Comment:
I'm in favor of moving this to a separate function or multiple delayed calls if necessary in order to avoid the dynamic function.